### ISSUE TITLE
Discussion about the way dimensions are checked

### ISSUE THREAD

**Issue by vsiles** [jerry73204/tch-typed-tensor#1]
Hi ! Nice work here.
I'm wondering if I understood correctly: your library adds static checks that are run and tested by the compiler at compile type, am I right ? Does it involves the type checker ?

**Comment by jerry73204** [comment]
Yep, type checks are done in compile time. Only code wrapping around libtorch involves runtime assertions to ensure same semantics of type and underlying implementation.

There's no extra type checker imported. It simply makes good use of Rust's type system. You _proof_ the correctness of a outcome type if the compiler can find an only way to construct that type (and within trait [bounds](https://doc.rust-lang.org/rust-by-example/generics/bounds.html)).

To be more exact, suppose you take a value of tensor by index. Only _indexing types_ within boundaries can be constructed. That's the way to prevent out-of-bound error.

**Comment by vsiles** [comment]
I see. I'll have to learn about typenum and type-freak a bit more before understanding everything, but it's really nice.
I'm afraid this is not an option for Python as the moment: they have literal types (Something like U32, N4, ...) but no operations on them yet. Something like that is planned for integer generics, but  it's not on anyone's radar afaik
